cgmanager.c

```diff
@@ -1114,6 +1131,253 @@ int list_children_main(void *parent, char *controller, const char *cgroup,
 }
 
+static int autoremove_entry_destroy(struct autoremove_entry *entry)
+{
+	nih_assert(entry != NULL);
+
+	nih_assert(entry->gpath != NULL);
+	nih_discard(entry->gpath);
+
+	nih_assert(entry->evpath != NULL);
+	nih_discard(entry->evpath);
+
+	nih_assert(entry->dirname != NULL);
+	nih_discard(entry->dirname);
+
+	if (entry->io != NULL)
+		nih_discard(entry->io);
+
+	nih_list_destroy(&entry->entry);
+
+	return 0;
+}
+
+/* if this function returns true then discontinue watching the events file */
+static bool autoremove_events_modified(struct autoremove_entry *entry)
+{
+	FILE *f;
+	char line[1024];
+	int pop_val = -1;
+	nih_local char *leafpath = NULL;
+
+	nih_assert(entry != NULL);
+	nih_assert(entry->evpath != NULL);
+
+	f = fopen(entry->evpath, "r");
+	if (!f) {
+		nih_error("%s: Cannot open %s in watcher: %s",
+			  __func__, entry->evpath, strerror(errno));
+		return false;
+	}
+
+	while (fgets(line, 1024, f) != NULL) {
+		int val;
+
+		if (sscanf(line, " populated %d", &val) != 1)
+			continue;
+
+		pop_val = !!val;
+		break;
+	}
+
+	fclose(f);
+
+	if (pop_val == -1) {
+		nih_error("%s: Cannot find or parse populated value in %s",
+			  __func__, entry->evpath);
+		return false;
+	} else if (pop_val == 1)
+		return false;
+
+	nih_assert(entry->gpath != NULL);
+	leafpath = NIH_MUST( nih_sprintf(NULL, "%s%s", entry->gpath, U_LEAF) );
+	if (rmdir(leafpath) < 0) {
+		if (errno == EBUSY)
+			return false;
+
+		nih_error("%s: Failed to remove %s: %s", __func__, leafpath,
+			  strerror(errno));
+	}
+
+	if (rmdir(entry->gpath) < 0) {
+		if (errno == EBUSY)
+			return false;
+
+		nih_error("%s: Failed to remove %s: %s", __func__, entry->gpath,
+			  strerror(errno));
+		return false;
+	}
+
+	nih_info(_("Removed %s as it was empty"), entry->gpath);
+	return true;
+}
+
+static void autoremove_inotify_read(void *data, NihIo *io, const char *buf,
+				    size_t len)
+{
+	struct autoremove_entry *entry = data;
+	struct inotify_event *event;
+
+	nih_assert(entry != NULL);
+	nih_assert(io != NULL);
+	nih_assert(buf != NULL);
+
+	bool should_exit = false;
+	while (len >= sizeof(*event)) {
+		size_t esize;
+
+		event = (struct inotify_event *)buf;
+		esize = sizeof(*event) + event->len;
+		if (len < esize)
+			break;
+
+		if (event->wd != entry->wd_cg &&
+		    event->wd != entry->wd_events) {
+			nih_error("%s: Got unknown watch descriptor %d (expected %d or %d) while watching %s",
+				  __func__, event->wd, entry->wd_cg,
+				  entry->wd_events, entry->gpath);
+			goto next;
+		}
+
+		if (event->mask & IN_IGNORED) {
+			nih_info(_("%s watch was removed"),
+				 entry->gpath);
+			should_exit = true;
+			goto next;
+		}
+
+		if (event->mask & IN_DELETE &&
+		    event->wd == entry->wd_cg)
+			do {
+				if (event->len < 1) {
+					nih_warn("got DELETE inotify event without object name");
+					break;
+				}
+
+				if (strcmp(event->name, entry->dirname) != 0)
+					break;
+
+				nih_info(_("%s was removed by somebody else"),
```

Comment:
Is this worth bailing the rest of the run on?

Reply:
Well, if the directory (cgroup) that we watch for emptiness in order to remove it get deleted by somebody else there is no point in continuing the watch (or its event processing).